### PR TITLE
Refactor significant portions of the codebase and refactor some land-atm handlers

### LINKED ISSUE #115: Refactor `phalf.py` and `pfull.py` to `handlers.yaml`
## Handlers to Refactor
 
Handlers that use CDAT -- uses `cdutil.vertical.reconstructPressureFromHybrid` (not in xarray natively)
- pfull.py
- phalf.py

### Other Info
For local package builds using `pip install .` or `python setup install`, the `setup.py` file is used.
The `install_requires` section of `setup.py` is supposed to list the required dependencies for the package that can be installed via pip (it doesn't have this section right now). This is most likely because `cdms2` and `cdutil` are only hosted on Anaconda and not pip. Therefore, building the local package using either command would result in these two dependencies to be missing the Conda environment.

Related lines:
- https://github.com/E3SM-Project/e3sm_to_cmip/blob/684416f69425ef52f6ec9c94203bbb5798d20c53/conda-env/dev.yml#L19-L22
- https://github.com/E3SM-Project/e3sm_to_cmip/blob/684416f69425ef52f6ec9c94203bbb5798d20c53/e3sm_to_cmip/cmor_handlers/pfull.py#L12-L17
- https://github.com/E3SM-Project/e3sm_to_cmip/blob/684416f69425ef52f6ec9c94203bbb5798d20c53/e3sm_to_cmip/cmor_handlers/phalf.py#L12-L17

### LINKED ISSUE #190: Refactor `lib.py` utilities to reduce code duplication and excessive nesting of function calls
Many utilities within `lib.py` contain duplicated code (conditionals, for loops, etc.). Refactoring these utilities can make it easier to extend functionality and fix bugs as needed. The script is lengthy and not well documented, thus is not easy to modify and maintain by developers.

Modules to refactor

- `handle_variables()`
  - [ ] Break up logic into smaller, private functions
  - [ ] Clean up and optimize logic
  - [ ] Move functions into `VarHandler` class (`cmorize` method wraps this function right now)
  - [ ] Remove call to `handle_simple()` if `simple=True` (this should be a separate function call)
- handle_simple()`
  - [ ] Reuse logic from `handle_variables()` via private functions

### LINKED ISSUE #218: Refactor `clcalipso.py` with `handlers.yaml`
`clcalipso.py` can be added directly to `handlers.py` without any additional code in `formulas.py`.

- Link: https://github.com/E3SM-Project/e3sm_to_cmip/blob/master/e3sm_to_cmip/cmor_handlers/vars/clcalipso.py

### PR DESCRIPTION
**Pull request by tomvothecoder**
## Overview
This PR was originally supposed only address #115, but I found the codebase difficult to work with so I decided to refactor it.

### Regression Testing

**TLDR: All of the datasets produced by this branch align with the `master` branch. The max relative differences between 5/109 datasets are insignificant and can be attributed to floating point rounding error produced by Xarray vs. CDAT, which is fine.**

---

Setup:
1. Generated branch datasets using [`scripts/branch-regression-testing/115-cdat-refactor-test/115-end-to-end-script.sh`](https://github.com/E3SM-Project/e3sm_to_cmip/blob/refactor/115-cdat/scripts/branch-regression-testing/115-cdat-refactor-test/115-end-to-end-script.sh) to compare against `master` branch datasets.
2. Compared datasets between branches using [`scripts/branch-regression/115-comparison-notebook.ipynb`](https://github.com/E3SM-Project/e3sm_to_cmip/blob/refactor/115-cdat/scripts/branch-regression-testing/115-cdat-refactor-test/115-comparison-notebook.ipynb)

Results:

- 104/109 are identical
- 4/109 are close and meet threshold of `rtol=1e-7` (`cl` `clw`, `cli`, `pfull`)
- 1/109 is not close (`mrso`)
  - `mrso` is fine (max relative difference of 6.726486e-07)
  ```python
	mrso
	-----
	MEAN
	    dev: 998.3673095703125, master: 998.3673095703125
	SUM
	    dev: 277705856.0, master: 277705856.0
	COMPARISON
	
	Not equal to tolerance rtol=1e-07, atol=1e-07
	
	Mismatched elements: 25556 / 777600 (3.29%)
	Max absolute difference: 0.00073242
	Max relative difference: 6.726486e-07
	 x: array([[[nan, nan, nan, ..., nan, nan, nan],
	        [nan, nan, nan, ..., nan, nan, nan],
	        [nan, nan, nan, ..., nan, nan, nan],...
	 y: array([[[nan, nan, nan, ..., nan, nan, nan],
	        [nan, nan, nan, ..., nan, nan, nan],
	        [nan, nan, nan, ..., nan, nan, nan],...
	```

### Core Changes

1. Refactor `lib.py` (now deleted)
   - Closes #190
   - Replace `run_parallel()` and `run_serial()` with `__main__.py` -> `E3SMtoCMIP._run_parallel()` and 
`E3SMtoCMIP._run_serial()` 
     - Updated var name `filepaths` to `vars_to_filepaths` for clarity (its a dict mapping var key to list of string paths)
     - Update args passed to handler method, remove unnecessary args
     - Add `try` and `except` statement for submitting `pool` jobs to maintain compatibility with MPAS variable handlers, which use different handler method arguments
   - Replace `handle_variables()`, `get_dimension_data()` and `load_axis()` with `VarHandler.cmorize()`
   - Delete `handle_simple()` -- _will be re-implemented from scratch in #130_
3. Update `handler.py`
   - Update `VarHandler.cmorize()`
      - Refactored significantly -- extracted logic into smaller, maintainable private methods 
      - Replaced `data` dictionary storing `xr.DataArrays`  with `ds` (`xr.Dataset` object)
	  - Distinguish CMOR usage with "cmor" string in function names and python variables
	  - Add support for hybrid sigma variables: `pfull`, `phalf`
   - Add `VarHandler._cmor_write()`
     - Add support for CMORizing fixed-time variables 
       - Separate PR for #217
     - If time dimension exists, CMORize the variable with all time and time bound values with a single call to `cmor.write()` instead of looping over each time value index and CMORizing each slice -- **this should improve performance and removes the `tqdm` progress bar.**
4. Update `handlers.yaml`
    - Add `phalf` and `pfull` entries 
      - Closes #115
      - Delete `phalf.py` and `pfull.py`
    - Add `clcalispo` entry 
      - Closes #218
      - Delete `clcalipso.py`
5. Update `_formulas.py`
   - Update all function argument types and return types from `np.ndarray` to `xr.DataArray`
   - Add formula functions for `pfull` and `phalf`
   - Add `convert_units()` function, which handles 1-to-1 unit conversions -- replaces `default.default_handlers.write_data()` 
6. Refactor `default.py` (now deleted)
   - Replaced by `VarHandler.cmorize()` and `_formulas.py.convert_units()`

### Clean Up Changes
7. Remove `cdms2` and `cdutil` from dependencies in `dev.yml` and `ci.yml`7
8.  Clean up legacy code in `clisccp.py` 
    - Separate PR for #218 
9. Add `Makefile` for easy access to commonly used commands (e.g., building and installing package)

### CODE REVIEW

File: e3sm_to_cmip/cmor_handlers/handler.py [hunk 1]
```diff
@@ -264,3 +259,46 @@ def _update_table_ref(self, freq: str, realm: str, cmip_tables_path: str):
         self.table = _get_table_for_non_monthly_freq(
             self.name, self.table, freq, realm, cmip_tables_path
         )
+
+    def _has_hybrid_sigma_levels(self):
+        if (
+            "ps" in self.raw_variables
+            and "hyai" in self.raw_variables
+            and "hybi" in self.raw_variables
+        ):
```
Comment (tomvothecoder) [inline e3sm_to_cmip/cmor_handlers/handler.py:268]:
```suggestion
        if (
            "PS" in self.raw_variables
            and "hyai" in self.raw_variables
            and "hybi" in self.raw_variables
            and "hyam" in self.raw_variables
            and "hybm" in self.raw_variables
        ):
```

**Comment by tomvothecoder** [comment]
Moved the section below out of the description to make the description shorter.

## Technical Debt in this Repo

I've noticed many Python anti-patterns and "not good" (_bad_) coding practices used in this repository. This has resulted in a lot of technical debt for new developers of this repo.
- A lot of copy-pasting code -- e.g., [legacy Python modules](https://github.com/E3SM-Project/e3sm_to_cmip/tree/master/e3sm_to_cmip/cmor_handlers/vars) for variable handlers, same logic in [`handle_variables()`](https://github.com/E3SM-Project/e3sm_to_cmip/blob/a98b69641b3b92214373f6f81e74aaca1c54d9a8/e3sm_to_cmip/lib.py#L439-L605)/[`handle_simple()`](https://github.com/E3SM-Project/e3sm_to_cmip/blob/a98b69641b3b92214373f6f81e74aaca1c54d9a8/e3sm_to_cmip/lib.py#L259-L421)/[`default_handler()`](https://github.com/E3SM-Project/e3sm_to_cmip/blob/a98b69641b3b92214373f6f81e74aaca1c54d9a8/e3sm_to_cmip/default.py#L9-L75)
- Functions using a large number of arguments, especially keyword arguments with default values ([example](https://github.com/E3SM-Project/e3sm_to_cmip/blob/a98b69641b3b92214373f6f81e74aaca1c54d9a8/e3sm_to_cmip/lib.py#L439-L455)) (and no API docstrings or type annotations for any of these functions).
- Using `**kwargs` to pass around function arguments ([example](https://github.com/E3SM-Project/e3sm_to_cmip/blob/a98b69641b3b92214373f6f81e74aaca1c54d9a8/e3sm_to_cmip/default.py#L9)) -- really fragile and hard to debug
- Extremely large functions that should be broken up into smaller, maintainable functions
- No unit testing excepts the ones I implemented :(
- Functions and data structures are unnecessarily nested in one another ([example](https://github.com/E3SM-Project/e3sm_to_cmip/blob/a98b69641b3b92214373f6f81e74aaca1c54d9a8/e3sm_to_cmip/default.py#L14-L59), [example](https://github.com/E3SM-Project/e3sm_to_cmip/blob/a98b69641b3b92214373f6f81e74aaca1c54d9a8/e3sm_to_cmip/lib.py#L656-L666))
- Excessive use of dictionaries to store handler information, variable and axes information, etc. ([example](https://github.com/E3SM-Project/e3sm_to_cmip/blob/a98b69641b3b92214373f6f81e74aaca1c54d9a8/e3sm_to_cmip/lib.py#L657-L718))

**Comment by tomvothecoder** [comment]
Hi @chengzhuzhang, this PR is finally ready for review! I performed thorough regression testing and everything checks out.

The PR description includes the core changes that you might want to focus on. Thanks!

File: e3sm_to_cmip/cmor_handlers/_formulas.py [hunk 2]
```diff
@@ -1,275 +1,394 @@
...
+    Notes
+    -----
+    This function is equivalent to `geocat.comp.interp_hybrid_to_pressure()`
+    and `cdutil.vertical.reconstructPressureFromHybrid()`.
+    """
+    result = ds["hyai"] * P0_VALUE + ds["hybi"] * ds["PS"]
+
+    # After Xarray broadcasting, the dimensions need to reordered from
+    # ["ilev", "time2", "lat", "lon"] to ["time2", "ilev", "lat", "lon"].
+    result = result.transpose("time2", "lev", "lat", "lon")
```
Comment (tomvothecoder) [inline e3sm_to_cmip/cmor_handlers/_formulas.py:265]:
```suggestion
    result = result.transpose("time2", "ilev", "lat", "lon")
```

File: e3sm_to_cmip/cmor_handlers/_formulas.py [hunk 3]
```diff
@@ -1,275 +1,394 @@
...
 
-def pr(data: Dict[str, np.ndarray], index: int) -> np.ndarray:
+    Notes
+    -----
+    This function is equivalent to `geocat.comp.interp_hybrid_to_pressure()`
+    and `cdutil.vertical.reconstructPressureFromHybrid()`.
+    """
+    result = ds["hyai"] * P0_VALUE + ds["hybi"] * ds["PS"]
+
+    # After Xarray broadcasting, the dimensions need to reordered from
```
Comment (tomvothecoder) [inline e3sm_to_cmip/cmor_handlers/_formulas.py:263]:
```suggestion
    # After Xarray broadcasting, the dimensions need to be reordered from
```

File: e3sm_to_cmip/cmor_handlers/_formulas.py [hunk 4]
```diff
@@ -1,275 +1,394 @@
...
+        The pfull pressure coordinates variable
+
+    Notes
+    -----
+    This function is equivalent to `geocat.comp.interp_hybrid_to_pressure()`
+    and `cdutil.vertical.reconstructPressureFromHybrid()`.
+    """
+    result = ds["hyam"] * P0_VALUE + ds["hybm"] * ds["PS"]
+
+    # After Xarray broadcasting, the dimensions need to reordered from
```
Comment (tomvothecoder) [inline e3sm_to_cmip/cmor_handlers/_formulas.py:227]:
```suggestion
    # After Xarray broadcasting, the dimensions need to be reordered from
```

File: e3sm_to_cmip/cmor_handlers/handler.py [hunk 5]
```diff
@@ -154,95 +177,544 @@ def to_dict(self) -> Dict[str, Any]:
...
+        if "time_bnds" in data_vars:
+            return "time_bnds"
+        elif "time_bounds" in data_vars:
+            return "time_bounds"
+
+        return None
+
+    def _reshape_single_time_bnd(self, ds):
+        # TODO: Reshape time bounds if it exists and has a length of 1: https://github.com/E3SM-Project/e3sm_to_cmip/blob/8c818fcb6d51cc0555f62b8058eee539b69a9579/e3sm_to_cmip/lib.py#L674-L678C42
+        pass
```
Comment (tomvothecoder) [inline e3sm_to_cmip/cmor_handlers/handler.py:440]: I added this TODO just in case it is needed. I didn't run into any issues where time bounds have a length of 1 and need to be reshaped like the old CDAT code linked in the comment.

File: e3sm_to_cmip/cmor_handlers/handler.py [hunk 6]
```diff
@@ -154,95 +177,544 @@ def to_dict(self) -> Dict[str, Any]:
...
+            For example, with this list:
+                [
+                    v2.LR.historical_0101.eam.h0.1850-01.nc,
+                    v2.LR.historical_0101.eam.h0.1850-02.nc,
+                    ...
+                ]
+            the value at index 0 is v2.LR.historical_0101.eam.h0.1850-01.nc.
+
+        time_dim : str | None
+            Whether or not the output CMIP variable has a time dimension.
```
Comment (tomvothecoder) [inline e3sm_to_cmip/cmor_handlers/handler.py:368]:
```suggestion
        vars_to_filepaths : Dict[str, List[str]]
            A dictionary mapping E3SM raw variables to a list of filepath(s).
        index : int
            The index representing the time range for the file.
            For example, with this list:
                [
                    v2.LR.historical_0101.eam.h0.1850-01.nc,
                    v2.LR.historical_0101.eam.h0.1850-02.nc,
                    ...
                ]
            the value at index 0 is v2.LR.historical_0101.eam.h0.1850-01.nc.
        time_dim : str | None
            Whether or not the output CMIP variable has a time dimension.
```

File: conda-env/ci.yml [hunk 7]
```diff
@@ -17,6 +15,7 @@ dependencies:
   - pyyaml
   - tqdm
   - xarray >=2022.02.0 # This version of Xarray drops support for Python 3.8.
+  - xcdat >=0.6.0
```
Comment (tomvothecoder) [inline conda-env/ci.yml:18]:
```suggestion
```

**Review by tomvothecoder — COMMENTED** [review]
Some last self-review

File: conda-env/ci.yml (hunk 7 above)
Comment (tomvothecoder) [reply conda-env/ci.yml:18]: xCDAT is actually not ever used.

File: e3sm_to_cmip/cmor_handlers/handler.py [hunk 8]
```diff
@@ -154,95 +177,544 @@ def to_dict(self) -> Dict[str, Any]:
...
+
+        Returns
+        -------
+        int
+            The ips CMOR axis id.
+        """
+        name = "ps"
+
+        # NOTE: This maintains the legacy name from pfull.py and phalf.py. I'm
+        # not sure why this is done and what the difference is with "ps".
```
Comment (chengzhuzhang) [inline e3sm_to_cmip/cmor_handlers/handler.py:600]: Thanks for the note. I don't have a good idea of this either.

**Review by chengzhuzhang — APPROVED** [review]
@tomvothecoder The PR looks good to me. Thanks for bringing in these nice changes and carefully tested the changes.  The code base is much cleaner now.

File: e3sm_to_cmip/cmor_handlers/handler.py [hunk 9]
```diff
@@ -154,95 +177,544 @@ def to_dict(self) -> Dict[str, Any]:
...
+        if self.levels is not None and self.levels["name"] == "sdepth":
+            ds["levgrnd_bnds"] = _formulas.LEVGRND_BNDS
+
+        return ds
+
+    def _get_time_bnds_key(self, data_vars: KeysView[Any]) -> str | None:
+        """Get the key for the time bounds.
+
+          * "atm" variables use "time_bnds"
+          * "ldn" variables use "time_bounds"
```
Comment (tomvothecoder) [inline e3sm_to_cmip/cmor_handlers/handler.py:419]:
```suggestion
          * "lnd" variables use "time_bounds"
```

File: e3sm_to_cmip/cmor_handlers/handler.py [hunk 10]
```diff
@@ -154,95 +177,544 @@ def to_dict(self) -> Dict[str, Any]:
...
+        if "time_bnds" in data_vars:
+            return "time_bnds"
+        elif "time_bounds" in data_vars:
+            return "time_bounds"
+
+        return None
+
+    def _reshape_single_time_bnd(self, ds):
+        # TODO: Reshape time bounds if it exists and has a length of 1: https://github.com/E3SM-Project/e3sm_to_cmip/blob/8c818fcb6d51cc0555f62b8058eee539b69a9579/e3sm_to_cmip/lib.py#L674-L678C42
+        pass
```
Comment (tomvothecoder) [inline e3sm_to_cmip/cmor_handlers/handler.py:440]:
```suggestion
```

Comment (tomvothecoder) [reply e3sm_to_cmip/cmor_handlers/handler.py:440]: Removing this code because I don't think it is needed. If a bug ever appears where time bounds are length 1, then we can add this feature.

**Comment by tomvothecoder** [comment]
Thanks @chengzhuzhang! I will merge this now.
